db/log_test.cc

```diff
@@ -774,6 +774,32 @@ TEST_P(LogTest, RecycleWithTimestampSize) {
   ASSERT_EQ("EOF", Read());
 }
 
+// Validates that `MaybeAddUserDefinedTimestampSizeRecord`` adds padding to the
+// tail of a block and switches to a new block, if there's not enough space for
+// the record.
+TEST_P(LogTest, TimestampSizeRecordPadding) {
+  bool recyclable_log = (std::get<0>(GetParam()) != 0);
+  const size_t header_size =
+      recyclable_log ? kRecyclableHeaderSize : kHeaderSize;
+  const size_t data_len = kBlockSize - 2 * header_size;
+
+  const auto first_str = BigString("foo", data_len);
```

Comment:
After `first_str`, there are `kBlockSize - (kBlockSize - 2 * header_size)` left in the block, a.k.a 2 * header_size:
1) for non recycleable format, that's 2 * (4 + 2 + 1) 
2) for recycleable format, that's 2 * (4 + 2 + 1 + 4)

It seems to me in both case, the remaining space in the block is sufficient to hold the ts_sz record which is:
1) for non recycleable format: 4 + 2 + 1 + 6
2) for recycleable format: 4 + 2 + 1 + 4 + 6

So it seems to me it won't invoke the logic added in this PR for: `if (leftover < header_size_ + (int)encoded.size()) `

If a ts_sz record can spill over multiple blocks, it seems to me we would need accomodating changes in log_reader.cc to handle any corresponding logic added into log_writer.cc

Maybe simply padding won't fully solve this issue, since it seems to me there is no other cases in log_writer/log_reader where one logical record spans over multiple blocks as multiple physical records without each physical record having its own header.

Reply:
Never mind about the ts_sz spanning over multiple blocks comment, I just realized this PR is not intended to target that.

Reply:
>After first_str, there are kBlockSize - (kBlockSize - 2 * header_size) left in the block, a.k.a 2 * header_size:

Hm, not exactly, it's actually `kBlockSize - (kBlockSize - header_size)==header_size`, since `Write(first_str)` writes a header + data, not only data itself into WAL.

Without the fix, this test actually fails on assert as I've written in the PR description